README.md
# Wtr - Well Typed Router 

Wtr - A HTTP request routing library for OCaml web applications.

[Documentation](https://lemaetech.co.uk/wtr/)&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; [Examples](https://github.com/lemaetech/wtr/tree/main/examples) &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; [Summer Demo](https://github.com/lemaetech/summer/blob/main/examples/echo_server.ml)

## Overview 

`wtr` is a HTTP request router for OCaml web applications.

- It is trie based so route matching is efficient and fast.
- Route handlers are type-checked during compilation.
- Supports matching and capturing URI path components, eg `/home/about/:int`.
- Supports matching and capturing URI query parameters, eg `/home/about?q=:int&q1=hello`.
- Supports matching HTTP methods, eg `GET, POST, PUT, HEAD, DELETE` etc.
- Supports converting captured URI components to OCaml and custom user defined data types.
- `%routes` ppx implements a DSL to allow you to easily specify HTTP uri. 
- combinatros based approach to specifying routes.

## Wtr Demo

```ocaml
# #require "wtr, wtr-ppx";;
```

```ocaml
module Fruit = struct
  type t = Apple | Orange | Pineapple

  let t : t Wtr.arg =
    Wtr.arg "fruit" (function
      | "apple" -> Some Apple
      | "orange" -> Some Orange
      | "pineapple" -> Some Pineapple
      | _ -> None )
end

let about_page = "about page"
let prod_page i = "Int page. number : " ^ string_of_int i
let float_page f = "Float page. number : " ^ string_of_float f
let contact_page nm num = "Contact. Hi, " ^ nm ^ ". Num " ^ string_of_int num
let product1 name id q = Format.sprintf "Product1 %s. Id: %d. q = %b" name id q
let product2 name id = Format.sprintf "Product2 %s. Id: %d." name id
;;
let fruit_page = function
  | Fruit.Apple -> "Apples are juicy!"
  | Orange -> "Orange is a citrus fruit."
  | Pineapple -> "Pineapple has scaly skin"

let faq category_id _ =
  let category_name =
    match category_id with
    | 1 -> "products"
    | 2 -> "insurance"
    | 3 -> "returns"
    | _ -> "unknown"
  in
  "FAQ page for category : " ^ category_name

let router =
  Wtr.router      
      [ {%routes| get,post,head,delete  ; /home/about/            |} about_page
      ; {%routes| head,delete           ; /home/:int/             |} prod_page
      ; {%routes| get,post              ; /home/:float/           |} float_page
      ; {%routes| get; /contact/*/:int                            |} contact_page
      ; {%routes| get; /product/:string?section=:int&q=:bool      |} product1
      ; {%routes| get; /product/:string?section=:int&q1=yes       |} product2
      ; {%routes| get; /fruit/:Fruit                              |} fruit_page
      ; {%routes| GET; /faq/:int/**                               |} faq ]

let p () = 
  Printf.printf "\n====Router Match Results====\n" ;
  [ Wtr.match' `GET "/home/100001.1/" router
  ; Wtr.match' `DELETE "/home/100001/" router
  ; Wtr.match' `GET "/home/about/" router
  ; Wtr.match' `GET "/product/dyson350?section=233&q=true" router
  ; Wtr.match' `GET "/product/dyson350?section=2&q=false" router
  ; Wtr.match' `GET "/product/dyson350?section=2&q1=yes" router
  ; Wtr.match' `GET "/product/dyson350?section=2&q1=no" router
  ; Wtr.match' `GET "/fruit/apple" router
  ; Wtr.match' `GET "/fruit/orange" router
  ; Wtr.match' `GET "/fruit/pineapple" router
  ; Wtr.match' `GET "/fruit/guava" router
  ; Wtr.match' `GET "/faq/1/" router
  ; Wtr.match' `GET "/faq/1/whatever" router
  ; Wtr.match' `GET "/faq/2/whateasdfasdfasdf" router ]
  |> List.iteri (fun i -> function
       | Some s -> Printf.printf "%3d: %s\n" (i + 1) s
       | None -> Printf.printf "%3d: None\n" (i + 1) );;
```

## Demo Output

```ocaml
# p () ;;
====Router Match Results====
  1: Float page. number : 100001.1
  2: Int page. number : 100001
  3: about page
  4: Product1 dyson350. Id: 233. q = true
  5: Product1 dyson350. Id: 2. q = false
  6: Product2 dyson350. Id: 2.
  7: None
  8: Apples are juicy!
  9: Orange is a citrus fruit.
 10: Pineapple has scaly skin
 11: None
 12: FAQ page for category : products
 13: FAQ page for category : products
 14: FAQ page for category : insurance
- : unit = ()
```


CHANGES.md
## v3.0.0 [2021-10-20] 

- BREAKING CHANGE A splat operator ** now expects a last string paramter
  in the handler. The parameter holds the remaining url matched by the splat operator
- BREAKING CHANGE create_decoder has been renamed to arg
- BREAKING CHANGE create has been renamed to router
- BREAKING CHANGE t type has been renamed to router
- BREAKING CHANGE a decoder has been renamed to a arg
- BREAKING CHANGE a uri has been renamed to a request_target
- FIX Fix matching query components in request target
- FIX Fix pretty printing of request_target
- NEW Add path, query and uri combinators to create and manipulate path, query and uri values
- NEW Add route, routes and pp_request_target functions
- NEW Add rest type and function rest_to_string
- CHANGE Refer splat as rest in the documentation.

## v2.0.0 2021-08-02 

- Convert assert based tests to ppx_expect tests
- Split README.md mdx test from tests
- BREAKING CHANGE ppx {%wtr| /home/about |} now produces a function
  expecting a handler rather than a uri. Demos and tests have been updated to
  reflect this change.
- BREAKING CHANGE remove functions pp_uri and >-.
- BREAKING CHANGE create function now mandates [route list list] to create
  a router. Hopefully this should not cause much breakage since the %wtr ppx 
  produces this value.
- NEW  %wtr ppx now allows specifying HTTP methods along with the uri route. A new 
  type meth has been added to represent all standard HTTP methods.
- NEW  Add Wtr.pp_route, Wtr.pp_method and Wtr.pp to pretty print a route, method 
  and a router respectively.

## v1.0.0 2021-05-11 UK

- First release.


otr/router.ml
(** Variable type. *)
module Var_ty = struct
  type _ t = ..

  type _ t += Int : int t | Float : float t | Bool : bool t | String : string t

  type (_, _) eq = Eq : ('a, 'a) eq

  let eq : type a b. a t -> b t -> (a, b) eq option =
   fun atid btid ->
    match (atid, btid) with
    | Int, Int -> Some Eq
    | Float, Float -> Some Eq
    | Bool, Bool -> Some Eq
    | String, String -> Some Eq
    | _ -> None
end

(** [('a, 'b) uri] represents a uniform resource identifier. The variant members
    describe the uri path component types. *)
type ('a, 'b) uri =
  | End : ('b, 'b) uri
  | Literal : string * ('a, 'b) uri -> ('a, 'b) uri
      (** Uri literal string component eg. 'home' in '/home' *)
  | Var : 'c var * ('a, 'b) uri -> ('c -> 'a, 'b) uri
      (** Uri variable component, i.e. the value is determined during runtime,
          eg. ':int' in '/home/:int' *)

and 'c var =
  { name : string (* name e.g. int, float, bool, string etc *)
  ; decode : string -> 'c option
  ; tid : 'c Var_ty.t
  }

let end_ : ('b, 'b) uri = End

let lit : string -> ('a, 'b) uri -> ('a, 'b) uri = fun s uri -> Literal (s, uri)

let var decode name tid uri = Var ({ decode; name; tid }, uri)

let string : ('a, 'b) uri -> (string -> 'a, 'b) uri =
 fun uri -> var (fun s -> Some s) "string" Var_ty.String uri

let int : ('a, 'b) uri -> (int -> 'a, 'b) uri =
 fun uri -> var int_of_string_opt "int" Var_ty.Int uri

let float : ('a, 'b) uri -> (float -> 'a, 'b) uri =
 fun uri -> var float_of_string_opt "float" Var_ty.Float uri

let bool : ('a, 'b) uri -> (bool -> 'a, 'b) uri =
 fun uri -> var bool_of_string_opt "bool" Var_ty.Bool uri

(** ['c route] is a uri and its handler. ['c] represents the value returned by
    the handler. *)
type 'c route = Route : ('a, 'c) uri * 'a -> 'c route

(** [p >- route_handler] creates a route from uri [p] and [route_handler]. *)
let ( >- ) : ('a, 'b) uri -> 'a -> 'b route = fun uri f -> Route (uri, f)

(** [uri_kind] Existential which encodes uri kind/type. *)
module Uri_kind = struct
  type t =
    | KLiteral : string -> t
    | KVar : 'c var -> t

  let equal a b =
    match (a, b) with
    | KLiteral lit', KLiteral lit -> String.equal lit lit'
    | KVar var', KVar var -> (
      match Var_ty.eq var.tid var'.tid with
      | Some Var_ty.Eq -> true
      | None -> false)
    | _ -> false

  (* [of_uri uri] converts [uri] to [kind list]. This is done to get around OCaml
     type inference issue when using [uri] type in the [add] function below. *)
  let rec of_uri : type a b. (a, b) uri -> t list = function
    | End -> []
    | Literal (lit, uri) -> KLiteral lit :: of_uri uri
    | Var (var, uri) -> KVar var :: of_uri uri
end

(** ['a t] is a node in a trie based router. *)
type 'a t =
  { route : 'a route option
  ; path : (Uri_kind.t * 'a t) list
  }

let update_path t path = { t with path }

let empty : 'a t = { route = None; path = [] }

let add (Route (uri, _) as route) (t : 'a t) =
  let rec loop t = function
    | [] -> { t with route = Some route }
    | uri_kind :: uri_kinds ->
      List.find_opt
        (fun (uri_kind', _) -> Uri_kind.equal uri_kind uri_kind')
        t.path
      |> (function
           | Some _ ->
             List.map
               (fun (uri_kind', t') ->
                 if Uri_kind.equal uri_kind uri_kind' then
                   (uri_kind', loop t' uri_kinds)
                 else
                   (uri_kind', t'))
               t.path
           | None -> (uri_kind, loop empty uri_kinds) :: t.path)
      |> update_path t
  in
  loop t (Uri_kind.of_uri uri)

type 'a t_compiled =
  { route : 'a route option
  ; path : (Uri_kind.t * 'a t_compiled) array
  }

let rec compile : 'a t -> 'a t_compiled =
 fun t ->
  { route = t.route
  ; path =
      List.rev t.path
      |> List.map (fun (uri_kind, t) -> (uri_kind, compile t))
      |> Array.of_list
  }

type decoded_value = D : 'c var * 'c -> decoded_value

let rec match' : 'b t_compiled -> string -> 'b option =
 fun t uri ->
  let rec loop t decoded_values = function
    | [] ->
      Option.map
        (fun (Route (uri, f)) ->
          exec_route_handler f (uri, List.rev decoded_values))
        t.route
    | uri_token :: uri_tokens ->
      let continue = ref true in
      let index = ref 0 in
      let matched_comp = ref None in
      while !continue && !index < Array.length t.path do
        let p = t.path.(!index) in
        match p with
        | KVar var, t' -> (
          match var.decode uri_token with
          | Some v ->
            matched_comp := Some (D (var, v) :: decoded_values, t');
            continue := false
          | None -> incr index)
        | KLiteral lit, t' when String.equal lit uri_token ->
          matched_comp := Some (decoded_values, t');
          continue := false
        | _ -> incr index
      done;
      Option.bind !matched_comp (fun (decoded_values, t') ->
          (loop [@tailcall]) t' decoded_values uri_tokens)
  in
  String.split_on_char '/' uri
  |> List.filter (fun tok -> not (String.equal "" tok))
  |> loop t []

and exec_route_handler : type a b. a -> (a, b) uri * decoded_value list -> b =
 fun f -> function
  | End, [] -> f
  | Literal (_, uri), decoded_values ->
    exec_route_handler f (uri, decoded_values)
  | Var ({ tid; _ }, uri), D ({ tid = tid'; _ }, v) :: decoded_values -> (
    match Var_ty.eq tid tid' with
    | Some Var_ty.Eq -> exec_route_handler (f v) (uri, decoded_values)
    | None -> assert false)
  | _, _ -> assert false

let r1 = string (int end_) >- fun (s : string) (i : int) -> s ^ string_of_int i

let r2 = lit "home" (lit "about" end_) >- "about"

let r3 = lit "home" (int end_) >- fun (i : int) -> "int " ^ string_of_int i

let r4 =
  lit "home" (float end_) >- fun (f : float) -> "float " ^ string_of_float f

let router = empty |> add r2 |> add r3 |> add r4 |> add r1

let router = compile router

let _m = match' router "/home/100001.1"

let _m1 = match' router "/home/100001"

let _m2 = match' router "/home/about"

(** This should give error (we added an extra () var in handler) but it doesn't.
    It only errors when adding to the router.*)
(* let r5 = *)
(* string (int end_) @-> fun (s : string) (i : int) () -> s ^ string_of_int i *)

(* |> add r5  *)
(* This errors *)
